address.md

```diff
@@ -0,0 +1,317 @@
+|------------|----------|---------------------|----------|
+| 'f' or 't' |    '3'   |  48 byte BLS PubKey |  4 bytes |
+                  base32[................................]
+```
+
+#### Payload
+
+The payload represents the data specified by the protocol. All payloads except the payload of the ID protocol are [base32](https://tools.ietf.org/html/rfc4648) encoded using the lowercase alphabet when seralized to their human readable format.
+
+#### Checksum
+
+Filecoin checksums are calculated over the address protocol and payload using blake2b-4. Checksums are base32 encoded and only added to an address when encoding to a string. Addresses following the ID Protocol do not have a checksum.
+
+
+#### Expected Methods
+
+All implementations in Filecoin must have methods for creating, encoding, and decoding addresses in addition to checksum creation and validation. The follwing is a golang version of the Address Interface:
+
+```go
+func New(protocol byte, payload []byte) Address
+
+type Address interface {
+	Encode(network Network, a Adress) string
+	Decode(s string) Address
+	Checksum(a Address) []byte
+	ValidateChecksum(a Address) bool
+}
+```
+##### New()
+
+New returns an Address for the specified protocol encapsulating corresponding payload. New fails for unknown protocols.
+
+```go
+func New(protocol byte, payload []byte) Address {
+	if protocol < SECP256K1 || protocol > BLS {
+		Fatal(ErrUnknownType)
+	}
+	return Address{
+		Protocol: protocol,
+		Payload:  payload,
+	}
+}
+```
+
+##### Encode()
+
+Software encoding a Filecoin address must:
+
+- produce an address encoded to a known network
+- produce an address encoded to a known protocol
+- produce an address with a valid checksum
+
+Encodes an Address as a string, prepending the network prefix, calculating the checksum, and encoding the payload and checksum to [base32](https://tools.ietf.org/html/rfc4648).
+
+```go
+func Encode(network string, a Address) string {
+	if network != "f" && network != "t" {
+		Fatal("Invalid Network")
+	}
+
+	switch a.Protocol {
+		case SECP256K1, Actor, BLS:
+			cksm := Checksum(a)
+			return network + a.Protocol + base32.Encode(a.Payload + cksm)
+		case ID:
+			return network + a.Protocol + base10.Encode(a.Payload)
+		default:
+ 			Fatal("invalid address protocol")
+	}
+}
+```
+
+##### Decode()
+
+Software decoding a Filecoin address must:
+* verify the network is a known network.
+* verify the protocol is a number of a known protocol.
+* verify the checksum is valid
+
+Decode an Address from a string by removing the network prefix, validating the address is of a know protocol, decoding the payload and checksum, and validating the checksum.
+
+```go
+func DecodeString(a string) Address {
+	if len(a) < 3 {
+		Fatal(ErrInvalidLength)
+	}
+
+	if a[0] != "f" && a[0] != "t" {
+		Fatal(ErrUnknownNetwork)
+	}
+
+	protocol := a[1]
+	raw := a[2:]
+	if protocol == ID {
+		return Address{
+			Protocol: protocol,
+			Payload:  base10.Decode(raw),
+		}
+	}
+
```

Comment:
```suggestion
raw = base32.Decode(raw)
```

Reply:
the checksum is base32 encoded as well, we have to decode the entire thing before slicing it off